Auto/IR/SMT.lean

```diff
@@ -415,29 +426,66 @@ section
   def hIn (e : ω) : TransM ω Bool := do
     return (← getH2lMap).contains e
 
-  /-- Used for e.g. bound variables -/
-  partial def disposableName : TransM ω String := do
+  def binderNamePrefixFromSort (sort : SSort) : String :=
+    match sort with
+    | SSort.bvar _ => "bvar"
+    | SSort.app (SIdent.symb s) _ => s.take 1
+    | SSort.app (SIdent.indexed s _) _ => s.take 1
+
+  /-- Used for e.g. bound variables. -/
+  partial def disposableName (sortHint : SSort): TransM ω String := do
     let l2hMap ← getL2hMap
     let idx ← getIdx
-    let currName := s!"smtd_{idx}"
+    let mut currName := s!"{binderNamePrefixFromSort sortHint}{idx}"
+    -- Try to avoid collisions with other identifiers
     if l2hMap.contains currName then
-      throwError "disposableName :: Unexpected error"
+      currName := s!"{currName}_{idx}"
+      if l2hMap.contains currName then
+        throwError "disposableName :: Unexpected error - binder disposable name already exists!"
     setIdx (idx + 1)
     return currName
 
+  def smtNameFromHint (nameHint : Option Expr) (default: String): TransM ω String := do
+    match nameHint with
+    | none => return default
+    | some nameHint => do
+      match nameHint with
+      | Expr.const name _ => return name.toString
+      | Expr.fvar fvarId =>
+        let lctx ← getLCtx
+        match lctx.find? fvarId with
+        | some localDecl => return localDecl.userName.toString
+        | none => return default
+      -- We might have different applications of the same polymorphic function;
+      -- to disambiguate, we use the argument names, not just the function names.
+      | Expr.app f args =>
+        let fname ← smtNameFromHint f default
+        let argname ← smtNameFromHint args "arg"
+        return s!"{fname}_{argname}"
+      | Expr.proj typeName idx _ => return s!"{typeName.toString}_proj_{idx}"
+      | _ =>
+        trace[auto.lamReif.printValuation] "smtNameFromHint :: Unexpected hint; saw {nameHint.ctorName} {nameHint} ({nameHint.dbgToString})"
+        return default
+
```

Comment:
The above ``smtNameFromHint`` will print out the entire expression, including all implicit arguments, which will result in very long strings in many cases. Consider the following example:
```
section

  variable (A : Type) (inst : Add A)

  axiom Rat_comm (a b : Rat) : a + b = b + a

  theorem Rat_comm' (a b : Rat) : a + b = b + a := by
    auto [Rat_comm]

end
```
What the pretty printer prints as ``@HAdd.hAdd Rat Rat Rat instHAdd a b`` (not showing implicit arguments) becomes ``HAdd.hAdd_Lean.Rat_Lean.Rat_Lean.Rat_instHAdd_Lean.Rat_Lean.Rat.instAddRat``. This issue will be more serious for ``MathLib4`` problems. For example, when using ``MathLib4``, what prints as ``⇑h`` can mean
```
@DFunLike.coe.{u_2 + 1, u_1 + 1, max (u_1 + 1) (u_2 + 1)}
    (@AddMonoidHom.{u_2, u_1} A₀ A₁
      (@AddMonoid.toAddZeroClass.{u_1} A₀
        (@SubNegMonoid.toAddMonoid.{u_1} A₀
          (@AddGroup.toSubNegMonoid.{u_1} A₀ (@AddCommGroup.toAddGroup.{u_1} A₀ inst✝⁵))))
      (@AddMonoid.toAddZeroClass.{u_2} A₁
        (@SubNegMonoid.toAddMonoid.{u_2} A₁
          (@AddGroup.toSubNegMonoid.{u_2} A₁ (@AddCommGroup.toAddGroup.{u_2} A₁ inst✝²)))))
    A₀ (fun (x : A₀) => A₁)
    (@AddMonoidHom.instFunLike.{u_2, u_1} A₀ A₁
      (@AddMonoid.toAddZeroClass.{u_1} A₀
        (@SubNegMonoid.toAddMonoid.{u_1} A₀
          (@AddGroup.toSubNegMonoid.{u_1} A₀ (@AddCommGroup.toAddGroup.{u_1} A₀ inst✝⁵))))
      (@AddMonoid.toAddZeroClass.{u_2} A₁
        (@SubNegMonoid.toAddMonoid.{u_2} A₁
          (@AddGroup.toSubNegMonoid.{u_2} A₁ (@AddCommGroup.toAddGroup.{u_2} A₁ inst✝²)))))
    h
```
In short, we might want to keep the smt Name closer to what the pretty printer prints. I think ``← e.toMessageData.format`` will do the job. Maybe append index when necessary to avoid collision.

Reply:
and replace non-unicode characters with underscore

Reply:
Sorry. To pretty print, use ``(← PrettyPrinter.delab e).raw.prettyPrint``.

Reply:
``← PrettyPrinter.formatTerm (← PrettyPrinter.delab e).raw`` might be even better